HTMLVideoElement.requestVideoFrameCallback can leak its captures if cancelVideoFrameCallback is not called.

rreno opened the pull request:
#### 5eb1d50bc03d3ec080f64562e5d5af4d1ef51b2a
<pre>
HTMLVideoElement.requestVideoFrameCallback can leak its captures if cancelVideoFrameCallback is not called.
<a href="https://rdar.apple.com/131184636">rdar://131184636</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=276267">https://bugs.webkit.org/show_bug.cgi?id=276267</a>

Reviewed by Youenn Fablet.

If HTMLVideoElement.requestVideoFrameCallback captures the document
object it will likely leak if cancelVideoFrameCallback is not called by
the web author. This change will clear pending and serviced video frame
callback lists in the HTMLVideoElement when ActiveDOMObject::stop is
called.

Additionally, I&apos;ve factored out some common code which I&apos;ve been using
in writing the Layout Tests for leak regression testing into a helper
file in LayoutTests/resources.

* LayoutTests/fast/html/request-video-frame-callback-does-not-leak-expected.txt: Added.
* LayoutTests/fast/html/request-video-frame-callback-does-not-leak.html: Added.
* LayoutTests/fast/html/resources/request-video-frame-callback.html: Added.
* LayoutTests/resources/document-leak-test.js: Added.
(createFrames):
(iframeForMessage):
(iframeLeaked):
(iframeSentMessage):
(runDocumentLeakTest):
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/html/HTMLVideoElement.cpp:
(WebCore::HTMLVideoElement::stop):
* Source/WebCore/html/HTMLVideoElement.h:

Canonical link: <a href="https://commits.webkit.org/280754@main">https://commits.webkit.org/280754@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/302231f53fb1b10c385e8fc82d1fa63f1ff335d0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57492 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36820 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9967 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61114 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7937 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/59620 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44453 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8125 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/46559 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5624 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59522 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34554 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49661 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27422 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31337 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6964 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6940 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/53305 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7236 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62793 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1405 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7329 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/53817 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1412 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49686 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/53912 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12718 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1201 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32649 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33734 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34819 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33480 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on LayoutTests/resources/document-leak-test.js:
```diff
@@ -0,0 +1,70 @@
+jsTestIsAsync = true;
+
+if (!(window.testRunner && window.internals)) {
```
```suggestion
if (!window.testRunner || !window.internals) {
```

youennf commented on LayoutTests/resources/document-leak-test.js:
```diff
@@ -0,0 +1,70 @@
+
+function runDocumentLeakTest(options)
+{
+    createFrames(options.framesToCreate);
```
Maybe comment why we create multiple frames.

rreno committed:
HTMLVideoElement.requestVideoFrameCallback can leak its captures if cancelVideoFrameCallback is not called.
rdar://131184636
https://bugs.webkit.org/show_bug.cgi?id=276267

Reviewed by Youenn Fablet.

If HTMLVideoElement.requestVideoFrameCallback captures the document
object it will likely leak if cancelVideoFrameCallback is not called by
the web author. This change will clear pending and serviced video frame
callback lists in the HTMLVideoElement when ActiveDOMObject::stop is
called.

Additionally, I've factored out some common code which I've been using
in writing the Layout Tests for leak regression testing into a helper
file in LayoutTests/resources.

* LayoutTests/fast/html/request-video-frame-callback-does-not-leak-expected.txt: Added.
* LayoutTests/fast/html/request-video-frame-callback-does-not-leak.html: Added.
* LayoutTests/fast/html/resources/request-video-frame-callback.html: Added.
* LayoutTests/resources/document-leak-test.js: Added.
(createFrames):
(iframeForMessage):
(iframeLeaked):
(iframeSentMessage):
(runDocumentLeakTest):
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/html/HTMLVideoElement.cpp:
(WebCore::HTMLVideoElement::stop):
* Source/WebCore/html/HTMLVideoElement.h:

Canonical link: https://commits.webkit.org/280754@main
